Tutorial: Port code for RemarkCommand

j-lum opened the pull request:
Now that the code base for AB3 has stabilized, let's port the removed tutorial from AB4 to AB3.
This should belong on its own branch. 

This commit does not contain the tutorial adoc. I'll be opening a PR to merge that into master.

j-lum committed:
Change RemarkCommand to throw an exception

Follow the convention and throw a CommandException instead of returning 
a CommandResult as the command did not fulfill its intended function.

Add RemarkCommandTest that will test that executeUndoableCommand()
throws an Exception.

Modify AddressBookParser to accept a RemarkCommand.

Add new test to AddressBookParserTest, which will test that typing
"remark" returns an instance of RemarkCommand.

j-lum committed:
Add Remark class to model

Add Remark to model component (copy from Address, remove the regex and
rename accordingly).

Add test for Remark (to test the equals method).

Modify RemarkCommand to now take in a Remark instead of a String.

j-lum committed:
Add a placeholder for remark in PersonCard GUI

Add a label to PersonListCard.fxml

Add FXML annotation in PersonCard to tie the variable to the actual
label.

j-lum committed:
Modify Person to support a Remark field

Add Person#getRemark()

Modify PersonBuilder to support Remark.

Modify AddCommand to create a Person with a blank Remark.

Modify EditCommand to pass Remarks on unchanged.

Modify SampleDataUtil to add remarks for the sample data.

Add withRemark() to PersonBuilder.

j-lum committed:
Add Remark field to JsonAdaptedPerson class

Update tests and test data to use the remark field.

j-lum committed:
Connect Remark field to PersonCard

j-lum committed:
Implement RemarkCommand execute() logic

Replace the error message with the actual logic to modify the remarks
of a person.

Update RemarkCommandTest to test that execute() logic works.

damithc commented:
I wonder if the initial steps can be even smaller. e.g., start with a simpler form of the command `remark` (not `remark 2 r/ good`). This will result in a smaller first commit.

On a related note, the current breakdown still feels like written by someone who knows the model solution, whereas the reader doesn't have that kind of expertise. Hence, the steps given may sound 'magical' to the reader (i.e., there is an empathy gap between the writer and the reader, which makes it harder for the reader to connect with the content). 
Keep in mind that we are not only training students to work with AB-3 code base, but also training them _how to use a strategy of 'tiny increments' when working in an unfamiliar code base_. The latter has more value but requires more effort in writing. For example, we should also explain the thought process in a way that can be generalized (`trace through a part that is similar to our target feature` instead of `trace through the delete command`)

j-lum commented:
From the code, we start from 
1. recognizing `remark` and ignoring everything else
1. recognizing the whole argument string
1. parsing the string to obtain more useful data

So I don't think the code needs to change, maybe we can emphasize the build up a bit more in the tutorial?

damithc commented:
> So I don't think the code needs to change, maybe we can emphasize the build up a bit more in the tutorial?

The first commit has more code than necessary. e.g.,
* any mention of prefix
* an elaborate help message
* test case

These are things an experienced AB programmer would include in the first step (because she knows these will be needed eventually) but not a newcomer.

j-lum commented:
I'm writing a new bridging section between the intro and the first bits to code to help the students case out the problem and break it down into logical components. From there we can re-evaluate how we want to structure the commits since the old tutorial (and its code) was meant to be the model answer.

j-lum commented:
Added a new intermediate step before `Logic: Teach the app to accept 'remark' but do nothing `.

The rationale was that students with little exposure to the code base isn't going to write standard-compliant code right from the code and we should ease them into the process.

damithc commented on src/main/java/seedu/address/logic/commands/RemarkCommand.java:
```diff
@@ -1,5 +1,8 @@
 package seedu.address.logic.commands;
 
+import static seedu.address.logic.parser.CliSyntax.PREFIX_REMARK;
```
Not needed for this step

damithc commented on src/main/java/seedu/address/logic/commands/RemarkCommand.java:
```diff
@@ -9,8 +12,18 @@
+            + "Parameters: INDEX (must be a positive integer) "
+            + PREFIX_REMARK + "[REMARK]\n"
+            + "Example: " + COMMAND_WORD + " 1 "
+            + PREFIX_REMARK + "Likes to swim.";
```
not needed for this step

damithc reviewed (changes requested):
2nd step needs to be simplified further as it contains things the reader will not see the need for. Plus, it needs to match the first step: currently, the first step creates a Result object and the second step switches to the Exception route.

damithc commented:
Is there a 1-1 mapping between the commits here and the sections in the tutorial? It's ideal if there is. In that case, section names and commit subjects should match closely.

damithc commented:
Let's move to the next iteration

j-lum committed:
Enhance RemarkCommand to accept and echo parameters

Modify RemarkCommand to take in an Index and String for remark, and
print those two parameters as the error message

Modify RemarkCommandTest to test the equals method

Add RemarkCommandParser that will know how to parse two arguments, one
index and one with prefix 'r/'

Add RemarkCommandParserTest that will test different boundary values
for RemarkCommandParser.

Modify AddressBookParser to use the newly implemented
RemarkCommandParser

Modify AddressBookParserTest to ensure that what the user input
generated the correct command.

j-lum committed:
Create a new `Remark` command

Add a minimum viable implementation of the `remark` command. This
implementation meant to be a bridging step between seeing a positive
result and doing things the right way.

Tests are not implemented because students are not expected to know how
to write them yet.

j-lum commented:
Not all commit titles are strict 1-1 matches but they are close enough with some minor differences in grammar. 

Going to shove this through for tutorial on week 6.
